rfcs/004 - Continuous Two-Way Synchronization RFC.md

```diff
@@ -0,0 +1,160 @@
+
+Each of these different "modes" have their own use-cases with various pros and cons, and they all seemed to be incompatible with each other. If the database is in one of the three automatic modes, there's no way to preview the changes that are going to happen. If the database is in read/write automatic mode, it won't restore critical infrastructure actually managed by IaSQL like the write-only automatic mode would, but write-only automatic mode prevents *any* other IaC tool from being used, so adoption would have to be all-or-nothing if that automatic restoration property was what brought someone to IaSQL in the first place.
+
+Recent work on the AWS `codebuild`, `codedeploy`, and `codepipeline` services has demonstrated that only "Two-Way" mode is actually viable: `codebuild`, for instance, accepts an arbitrary, user-defined YAML file describing the process to build code from some place and store it in some other place. IaSQL cannot know before it triggers `codebuild` what it will do, and so during the `iasql_apply` call suddenly there will exist resources, such as a new ECR image, that are not in the database which `iasql_apply` will want to remove because that's not part of the "desired" state as far as the engine is concerned, but *is* the true desired outcome from the user.
+
+Attempting to `iasql_sync` in the middle of that `iasql_apply` could cause IaSQL to drop other operations it is supposed to be performing, freezing their cloud into some unexpected intermediate state. And it is impossible to "parse" the YAML to determine what resources are going to be created. Since the YAML allows arbitrary shell script execution, the script could do something like `if (Math.random() > 0.5) { writeToS3(); } else { writeToECR(); }` or even `curl https://raw.githubusercontent.com/company/project/some_binary_that_calls_aws_apis; chmod +x some_binary_that_calls_aws_apis; ./some_binary_that_calls_aws_apis;` and there's no good, general purpose way to know what the side-effects of `codebuild` will be.
+
+However, it is possible for "Two-Way" mode to emulate the other three in a way that is also mostly-backwards-compatible with the current manual mode. Furthermore, a singular IaSQL database would be able to have different "parts" of the database emulating different modes at the same time.
+
+To accomplish this, let's assume under normal operations, there is a singular apply/sync-combined command to perform the "Two-Way"-style action. It works similarly to the current sync and apply, but also uses the audit log to decide for each record which behavior to follow. This piece can be operated in a "manual" style, invoked explicitly, but we'll instead invoke it through a cron-like interval, every few minutes, to make the automatic mode. In this way, there is no special syntax for users to work with normally. They just `SELECT/INSERT/UPDATE/DELETE` records they see and eventually this is reflected in their cloud account.
```

Comment:
I think there are no implications to migration usage and in fact it simplifies the UX for it, but it might be worth a mention

Reply:
Addressed.